# Preface to {{< var ay >}} BM214 Workshop 3: Enumeration of Bacteria and Bacteriophage {.unnumbered}

Welcome to the BM214 (Being a Biomolecular Scientist 2) workshop on enumeration of bacteria and bacteriophage, for {{< var ay >}}.

This year is the first presentation of the workshop material in this format, and we would be very grateful to hear feedback [by email](mailto:{{< var admin.email >}}) or through the [GitHub repository Issues page]({{< var github.issues >}}).

## Overview

This workshop asks you to work through some computational exercises that use the sequence, growth, and expression data generated in BM241 laboratories 1-3.

- In the first exercise, you will use a 16S sequence amplified by PCR from your unknown bacterium to query a public sequence database and identify your organism.
- In exercise two, you will use `R` and `ggplot2` to plot and interpret absorbance data from your yeast growth experiment.
- In the third exercise, you will again use `R` and `ggplot2` to plot and interpret absorbance ratio data from your candidate reporter experiment.

::: { .callout-tip }
We also provide an ["`R` playground"](playground.qmd) if you would like to get further practice interacting with and plotting data in `R`.
:::

::: { .callout-warning title="Important Note" }
There is new material in this workshop that is not covered in lectures, and this material is examinable. Please take care to reach the text in the expandable callout boxes, as well as that for the workshop activities, to be sure you have understood the topic and obtain full value from the exercises.
:::

## Learning Objectives

By the end of this workshop, students will be able to:

- explain that 16S sequences are commonly used for bacterial identification and taxonomic assignment
- describe and use two online tools for 16S sequence analysis: `BLAST` and `silva`
- interpret the output of `BLAST` and `silva` and positively identify an organism, or rule out an identification
- plot data from a microbial growth curve using `R` and `ggplot2`
- fit a curve to microbial growth data using `R` and `ggplot2`
- plot data from a reporter assay using `R` and `ggplot2`
- adjust and customise a figure using `R` and `ggplot2`

## Assessment

The workshop is not formally assessed although, as noted above, all the material contained in the workshop is examinable. There are three exercises for you to work thorough: one using online bioinformatics web services to identify an organism from its 16S sequence; and two using `R` to visualise and interpret experimental data.

::: { .callout-important }
There are two formative quizzes covering the exercises in the workshop, and these can be found on the workshop [MyPlace page](https://classes.myplace.strath.ac.uk/course/view.php?id=21632). You should complete these as part of the workshop.

- [MyPlace formative quiz 1]({{< var myplace.quiz1 >}}) - parts 2 and 3
- [MyPlace formative quiz 2]({{< var myplace.quiz2 >}}) - for the whole workshop
:::

